### ISSUE TITLE
Change cache folder to being script-local

### ISSUE THREAD

**Issue by ChrSteinert** [scriptcs/scriptcs#1086]
When executing scriptcs with the -cache flag a hidden cache directory gets created. This directory, however, is created at the working directory, not at the script directory. I have a collection of scripts in a directory and run them via powershell to manipulate files in any other directory. Using the -cache flag, the cache gets created in my current working directory. This is a) not useful b) cluttering my drive with cache folders.
So, either -cache could take a target directory for cache files or the cache directory could be implicitly created in the scripts directory (which I find more intuitive).

**Comment by adamralph** [comment]
@ChrSteinert thanks for raising this.

The default behaviour should almost certainly be in the scripts directory.

**Comment by adamralph** [comment]
@ChrSteinert I propose we use this issue to represent changing the behaviour to create the cache folder in the script directory. If we still feel that the directory needs to be controllable we can raise a separate issue for that. What do you think?

**Comment by ChrSteinert** [comment]
I'm perfectly fine with that. This is what I would prefer anyway. Just wanted to give some alternative seeing conversion around more powerful arguments going on here in the issues.

**Comment by ChrSteinert** [comment]
I'll take it. Am close to a solution. Tracked down where you guys set the dirs. So shouldn't be a big problem.

**Comment by ChrSteinert** [comment]
I'll look into it further. Let's see what can be done. This however was the easiest way ;)

**Comment by ChrSteinert** [comment]
So I found a non breaking solution. This, however, feels a bit dirty. The cache dir is now set after init of the ScriptExecutor class. This seems to be the only way, currently, to get the correct dir into this class without changing the interfaces. Have a look: https://github.com/ChrSteinert/scriptcs/commit/3652b87ce6f0c3f81d2491693203d293d42ba414

**Comment by adamralph** [comment]
@ChrSteinert I see what you mean. Whilst it's tempting as a one liner, it is indeed dirty and it really ought to be in `IScriptExecutor.Initialize()`.

@scriptcs/core how do we feel about breaking `IScriptExecutor.Initialize` and introducing a path argument which would be the current working dir when running a bare REPL and the script dir when running a script?

Note that the problem is also more general since we are also combining the current working dir instead of the script dir with `FileSystem.BinFolder` and assigning that to `ScriptExecutor.ScriptEngine.BaseDirectory`.

Another option could be to add this path argument into the `ScriptExecutor` constructor but it would be the only primitive param so I guess it would screw up our IoC.

**Comment by ChrSteinert** [comment]
Tough call. I'd suggest adding a CachDirectory property to the IFileSystem interface. This, while not non-breaking, would not break DI.
On side note. I personally felt it weird that the cache dir etc. is initialized in the ScriptExecutor class. I'm not judging this design, just being curious. Shouldn't this be somewhere else? The cache dir never gets used if a REPL is started and no script is passed as argument, but gets initialized anyway. Might this be target of refactoring in the future?

**Comment by adamralph** [comment]
The piece of knowledge we really want passed around is the script location, the location of the cache directory is just one usage of it.

You make a good point about the design. It sounds like there really ought to be an `Executor` base class with both `ScriptExecutor` and `Repl` inheriting from it. The problem is that any changes like these are breaking given our current approach of exposing our internals to the world as NuGet packages (something to be discussed elsewhere).

**Comment by khellang** [comment]
> It sounds like there really ought to be an Executor base class with both ScriptExecutor and Repl inheriting from it.

#435 :wink:

**Comment by ChrSteinert** [comment]
So, how to go onwards from here?

**Comment by filipw** [comment]
i think you can obsolete the existing `Initialize` method and add a new `Initialize`. There instead of rather meaningless flat parameters `string scriptDir, IEnumerable<string> paths, IEnumerable<IScriptPack> scriptPacks, params string[] scriptArgs` you could use a new `ScriptExecutionContext` class where you can wrap all these information. I can imagine there will be a need to pass more info into this context in the future too

Then in 2 releases or so, we can remove the old `Initialize` completely

**Comment by adamralph** [comment]
If we are going to break the interface then obsoleting for a version or two is probably a good idea.

However, since it's an interface, the change will still eventually be breaking. There's no way for an implementer to switch away from the obsolete to avoid breakage.

**Comment by glennblock** [comment]
A non-breaking alternative would be to use an environment variable.

**Comment by ChrSteinert** [comment]
IMO this smells like a god object thingy. I don't feel, especially an initialization process should be driven by data made accessible globally for the sake of this process. Then the one liner posted above is the way to go. This, however, is not extendable nor gives me the feeling of good design.
I'm currently working on the new InitContext like suggested above. Let's see where this goes, shall we?

**Comment by glennblock** [comment]
It feels like the right place for this is in FileSystem as a property and
not on the executor. This is how we have handled all path related things.

**Comment by glennblock** [comment]
The executor can easily have access to the filesystem snd check for the
property to get the cache folder.

**Comment by khellang** [comment]
I'd consider this within the "scripting domain", not the "file system domain". I'm not sure `IFileSystem` is the correct place for this. :-1: It would a breaking change anyway.

**Comment by glennblock** [comment]
We already have a DllCacheFolder property on IFileSystem similar to a
global folder so there is precedent. We have used the filesystem for these
sort of things. Yes it is a breaking change, everything we are discussing
is a breaking change.

**Comment by filipw** [comment]
yeah I agree with @khellang, to me it is an attribute of a context of an executing script.
The `DllCacheFolder` on the file system was intended to be just a constant identifying the folder name (just like other properties there, identifying packages file name or bin folder name).

**Comment by ChrSteinert** [comment]
I've taken a look inside the `IFileSystem` interface and according class.The to be executed script seems not to be in there (is it?). So the relative cache path is not of much use to this. The path to the script would have to be included. Is that really FileSystem concern?

Maybe the best solution, for now, would be to take the workaround in setting the cache folder in the `ExecuteScriptCommand` method. Shaping a new context class would need further input of the team, I think (but I'd be happy to help ;) ).

**Comment by ChrSteinert** [comment]
For the context-style approach: This could be something like this: https://github.com/ChrSteinert/scriptcs/commit/10cacb338d77cff7960d72ec3a5c478f5bc176da

**Comment by glennblock** [comment]
No the executed script is not there. What could be there is a property that
describes the behavior ie an enum that specifies working vs script local.
The executor then use this.

**Comment by glennblock** [comment]
Thinking about it more I like the context class idea as it will allow
extensibility in the future without breaking changes.

Aside from a typo on ReferencePaths this looks like it is in the right
direction.

**Comment by ChrSteinert** [comment]
Whoops. Nowhere near ready anyway. Saw some problems with unaligned usings and suff. But I'd rather wait for further input before proceeding with this. Will happily build something like this if you guys are interested.

**Comment by khellang** [comment]
The context thingy that @filipw mentioned is exactly what I was trying to do with #225 (before I gave up) by piping `IScriptEnvironment` through layers and layers of abstractions. It's not really clear who's got which responsibilities right now.

**Comment by glennblock** [comment]
That was why I was trying to use filesystem as it already exists. Not
perfect, but on the other hand this has not been a common request and it
only affects hosters

**Comment by glennblock** [comment]
I agree we need to level up and look at our design as we've grown
dramatically from the early days and it is showing its signs of age.

**Comment by glennblock** [comment]
Talking about this specific change is not quite the same as what you are
doing though as this context class is specifically scoped to the executor
and is not passed through tons of levels right?

**Comment by glennblock** [comment]
You meaning @khellang

**Comment by khellang** [comment]
Sure, but they serve much of the same purpose. The current executing script could easily be a part or the script environment object as well.

**Comment by glennblock** [comment]
I see what you mean.

Looking at this a little more, I am not sure Initialize is the right place
for this.

The proposal is the cache folder sits next to the script. If that is the
case then it seems like the place to pass it is in the Execute method as
that can get called over and over for diff scripts. That would also gel
well with capturing the root script which is passed in to Execute.

**Comment by ChrSteinert** [comment]
Is that so? `Execute` seems to be called only once for both instances (script and REPL). I might have overlooked something there but this is as I saw it. What I have not tested (but will soon) is what happens if you have a `#load` directive in the passed script. Where will those caches be? Where should they be? Maybe this whole thing is due for an overhaul, I'm afraid.

**Comment by filipw** [comment]
`Execute` is called only once when you use `scriptcs.exe`. `#load` is not creating new caches, it's all cached in the same final assebmly.

What @glennblock meant is that if you are hosting scriptcs in your own .NET application (after all `scriptcs.exe` is just a host application) you might be calling `Execute` multiple times, but not reinitialize over and over.

**Comment by glennblock** [comment]
If someone hosts scriptcs they can use the same executor over and over. Yes
in the REPL you are correct.

**Comment by ChrSteinert** [comment]
Ah, have not considered this. Hosting this seems to be a not so well documented feature, isn't it? Never came to my mind at least. I see the complexity increasing ...

**Comment by filipw** [comment]
here is a good hosting example: https://github.com/scriptcs-contrib/scriptcs-owin-host-iis
here is another from NDC Oslo 2014: https://github.com/filipw/sample-scriptcs-webhost

**Comment by glennblock** [comment]
It is not well documented other than blog
posts. Here is a good blog post on the topic:
http://madstt.dk/endeavours-with-scriptcs-hosting/

**Comment by ChrSteinert** [comment]
Wow this is awesome stuff. Thanks!

**Comment by glennblock** [comment]
Though it is not well documented (yes we need to change that) it is
absolutely a very intentional part of our design.

**Comment by glennblock** [comment]
As Filip mentioned, caching does not relate to #load. I think it will be
safe to introduce the cache folder at the point of script execution based
on this. We do support today loading a script into the REPL using the -R
param and providing a script name, but I think it is fine if caching does
not work in that context.

**Comment by ChrSteinert** [comment]
I'm currently writing up some solutions based on the different approaches mentioned here like the `FileSystem` class as the information holder, the new `InitContext` class (see this branch https://github.com/ChrSteinert/scriptcs/tree/InitContext) and the hotfix (see this branch https://github.com/ChrSteinert/scriptcs/tree/cacheDirHotfix). 
I'm having trouble with the FileSystem approach. The `FileSystem` class is injection via DI. So where would be the best location to set the directory of the current executed script? Any advice?
